EDU-14302: Managing lazy loading sections on Product Listing Pages (#1723)
vtexdocs/dev-portal-content: docs/faststore/docs/managing-performance/managing-lazy-loading-sections-on-plp.mdx

While lazy loading improves performance, it can harm SEO. Search engine crawlers often do not execute JavaScript or wait for dynamically loaded content. As a result, critical sections like product grids or category pages may not be indexed. For instance, if a user searches for "gym clothes," your product listings might not appear in search results if they are not included in the initial HTML.

In this guide, learn how to optimize SEO by controlling lazy loading for sections in FastStore.

## Before you begin

Make sure the `@faststore/cli` package is updated to the latest version. If you need to update it, follow the instructions on [Updating the '@fastore/cli' package version](https://developers.vtex.com/docs/guides/faststore/project-structure-updating-the-cli-package-version)

## Instructions

### Step 1: Configuring SEO for Product Listing Pages (PLP)

> Configure the title and description templates for product listing pages (PLP) in the discovery.config.js file to optimize your store's SEO and allow dynamically generating metadata based on the collection name.

Reviewer: suggestion: since the title template would also affect the other types of PLPs, I believe it would be good to not talk only about "collection" -- but I'm not sure on how to change this text hahah what do you think? Also, if the category/brand doesn't have a description configured on Catalog, the description template would be used. 
```suggestion
Configure the title and description templates for product listing pages (PLP) in the `discovery.config.js` file to optimize your store's SEO and allow dynamically generating metadata based on the collection/category/subcategory/brand name.
```

1. Open your FastStore project in a code editor of your preference.
2. Go to the `discovey.config.js` file and in the `seo` section add the following:

   ```discovery.config.js
   seo: {
       plp: {
       titleTemplate: '%s | FastStore PLP',
       descriptionTemplate: '%s products on FastStore Product Listing Page',
       },
   },
   ```

>    > ⚠ Change the values in `titleTemplate` and `descriptionTemplate` according to your collection’s name.

Reviewer: ```suggestion
   > ⚠ Change the values in `titleTemplate` and `descriptionTemplate` according to your needs. The `%s` should be used where you would like the collection name to appear. 
```

### Step 2: Controlling lazy loading for custom sections

By default, sections in FastStore are lazy-loaded for performance reasons. However, some sections (e.g., those containing H1 headers) may need to appear in the initial HTML for SEO purposes. To achieve this, you can configure a section to skip lazy loading.

1. Open the `cms/faststore/sections.json` file in your project.

> 2. Locate the custom section, add the `skipLazyLoadingSection` property, and set it to `true`. In the following example, we created new custom section called `CustomCollectionName` that displays the store collection name:

Reviewer: suggestion: we could also link the doc that explains how to create a new custom section
suggestion 2: make it clear that we'll also show an example of the section implementation. I don't know the best way to say this, but something like `First we'll add this section to the sections.json file, and then, on step 3 we'll create this section's component` just to link it better to the next step - but I think step 2 and 3 are so linked that it should be on a single "step section" of the doc 🤔 
```suggestion
2. Locate the custom section, add the `skipLazyLoadingSection` property, and set it to `true`. In the following example, we created a new custom section called `CustomPLPName` that displays the PLP (category/subcategory/collection/brand) name:
```

```sections.json
…
{
  "name": "CustomCollectionName",
  "schema": {
    "title": "Collection Name",
    "description": "Custom H1 configuration",
    "type": "object",
    "required": ["skipLazyLoadingSection"],
    "properties": {
      "skipLazyLoadingSection": {
        "title": "Skip lazy loading",
        "type": "boolean",
        "default": true
      }
    }
  }
}
```

> ### Step 3: Accessing collection data in a custom section
> 
> To display the collection name or other SEO-related data in a custom section, use the [`usePLP()`](https://developers.vtex.com/docs/guides/faststore/api-extensions-consuming-api-extensions#consuming-api-extensions-data-from-custom-sections) hook in the section’s file:

Reviewer: ```suggestion
### Step 3: Accessing collection data in a custom section

To display data related to the PLP in a custom section, use the [`usePLP()`](https://developers.vtex.com/docs/guides/faststore/api-extensions-consuming-api-extensions#consuming-api-extensions-data-from-custom-sections) hook in the section’s file:
```

```CallToAction.tsx
import { usePLP } from '@faststore/core'

export default function CustomCollectionName() {
  const context = usePLP()

  return (
    <section>
      <h1>Testing - {context.data?.collection?.seo?.title ?? 'PLP'}!</h1>
    </section>
  )
}
```

### Sending the changes to the Headless CMS

1. Open the terminal and run `faststore cms-sync` to sync your local changes with the Headless CMS.
2. Go to the VTEX Admin, access **Storefront > Headless CMS**.
3. Access the **Product List Page** Content Type.
4. Add the custom section you created. At the end of the section, you’ll notice the **Skip lazy loading** option.

   ![skip-lazy-loading-ui-option](https://vtexhelp.vtexassets.com/assets/docs/src/skip-lazy-loading-option___32bda38cafbf728774a30ad9ada8be01.png)

5. Click `Save` and publish the changes you made.
6. Click `Preview`.

> 7. Search for a collection name to see your new section. For example, if you have `Just arrived` collection added to the store slug, for example: `https://mystore.vtex.app/just-arrived`. You’ll be able to see something similar to the following:

Reviewer: ```suggestion
7. Search for a collection name to see your new section. For example, if you have a `Just arrived` collection, add its slug to the URL, for example: `https://mystore.vtex.app/just-arrived`. You’ll be able to see something similar to the following:
```

    ![custom-section-ui](https://vtexhelp.vtexassets.com/assets/docs/src/custom-section___48bf7d70d678b4fb4ddae39aa0cf14cb.png)

### Checking if the custom section is visible to search engines

To make sure that the custom section is present in the initial HTML response, follow these steps:

1. In the browser, right-click anywhere on the page and select **Inspect** from the context menu.

> 2. In the **Developer Tools** panel, click on the **Network tab**. Ensure that the **All filter** is selected.

Reviewer: You can filter by "Doc"

While lazy loading improves performance, it can harm SEO. Search engine crawlers often do not execute JavaScript or wait for dynamically loaded content. As a result, critical sections like product grids on category pages may not be indexed. For instance, if a user accesses the "gym clothes" collection, its product listing might not appear in search results if it's not included in the initial HTML.

In this guide, learn how to optimize SEO by controlling lazy loading for sections in FastStore.

## Before you begin

Make sure the `@faststore/cli` package is updated to the latest version. If you need to update it, follow the instructions on [Updating the '@fastore/cli' package version](https://developers.vtex.com/docs/guides/faststore/project-structure-updating-the-cli-package-version).

## Understand how SEO metadata is determined

FastStore follows a specific priority order when generating SEO metadata for PLP pages:

<details>
<summary>Section title</summary>

```mermaid
graph LR

> A[Checking the Catalog for product's title] --> B{Is there a title in the Catalog?};

Author: Thanks, Lari! We should go with 'Checks,' as in 'The system checks...'.

B -- Yes --> C[Use the title from the Catalog];

> B -- No --> D[Check 'discovery.config.js' file];

Author: Something like the following?
![Screenshot_389](https://github.com/user-attachments/assets/aaeca719-a53f-4d81-bbf1-20999a3882a9)

### Step 4: Checking if the custom section is visible to search engines

To make sure that the custom section is present in the initial HTML response, follow these steps:

1. In the browser, right-click anywhere on the page and select **Inspect** from the context menu.
2. In the **Developer Tools** panel, click on the **Network tab**. Ensure that the **All filter** is selected.

  > ℹ To focus only on the main HTML document, you can filter by **Doc** in the filter bar. This will show only the initial HTML request, making it easier to inspect the page's source code.

3. Reload the page while the Developer Tools are open. This will capture all network requests made during the page load.
4. In the **Name** column of the **Network** tab, look for the request corresponding to the page you're inspecting In our example `just-arrived`.
5. Click the **Response** tab in this panel.
6. If the section is present in the initial HTML response, you will see the corresponding HTML code in the **Response** tab. For example:

   ![developer-tools-panel](https://vtexhelp.vtexassets.com/assets/docs/src/developer-tools-panel___9240a6f4b36ce8151f6d346e78674f2b.png)

> If the section is present in the initial HTML, as in the example above, it means it is not lazy-loaded and is visible to search engines.

Reviewer: Perfect!

---

> title: "Managing lazy loading sections on Product Listing Pages (PLPs)"

Reviewer 2: ```suggestion
title: "Optimizing SEO by managing lazy loading on Product Listing Pages (PLPs)"
```

Reviewer 2: O objetivo é principalmente melhorar questões de SEO em PLPs. Acho que isso deve estar indicado no título da doc

While lazy loading improves performance, it can harm SEO. Search engine crawlers often do not execute JavaScript or wait for dynamically loaded content. As a result, critical sections like product grids on category pages may not be indexed. For instance, if a user accesses the "gym clothes" collection, its product listing might not appear in search results if it's not included in the initial HTML.

In this guide, learn how to optimize SEO by controlling lazy loading for sections in FastStore.

> ⚠ This feature may increase initial page load times, potentially impacting performance and user experience. Use it only when necessary.

## Before you begin

Make sure the `@faststore/cli` package is updated to the latest version. If you need to update it, follow the instructions on [Updating the '@fastore/cli' package version](https://developers.vtex.com/docs/guides/faststore/project-structure-updating-the-cli-package-version).

## Understand how SEO metadata is determined

> FastStore follows a specific priority order when generating SEO metadata for PLP pages:

Reviewer 2: ```suggestion
FastStore follows a structured approach to generating SEO metadata for PLPs, prioritizing available catalog data before falling back to predefined templates in `discovery.config.js`.
```

E -- Yes --> F[Uses the name from the Catalog];
E -- No --> G[Checks 'discovery.config.js' file];
G --> H[Uses general SEO title];
C --> I[Displays Title];
F --> I;
H --> I;
```

</details>

<details>
<summary>Title template</summary>

```mermaid
graph LR
A[Checks for a template configured in 'discovery.config.js' in the 'seo' section] --> B{Is there a title template?};
B -- Yes --> C[Displays Title using the template];
B -- No --> D[Uses the generic title template from the 'seo' object in 'discovery.config.js'.];
D --> C;
```

</details>

<details>
<summary>Description template</summary>

```mermaid
graph LR
A[Checks the Catalog for the PLP's description] --> B{Is there a description in the Catalog?};
B -- Yes --> C[Uses the description from the Catalog];
B -- No --> D[Checks the PLP description template in the 'plp' property of 'discovery.config.js'];
D --> E{Is there a PLP description template?};
E -- Yes --> F[Uses the PLP description template using the PLP title in it];
E -- No --> G[Uses the generic description template in the 'seo' object of 'discovery.config.js'];
C --> H[Displays Description];
F --> H;
G --> H;
```

</details>

## Instructions

> ### Step 1: Configuring SEO for Product Listing Pages (PLP)

Reviewer 2: de repente isso deveria estar dentro de before you begin. achei que ficou meio deslocado esse conteúdo aqui

> Configure the title and description templates for product listing pages (PLP) in the [`discovery.config.js`](https://developers.vtex.com/docs/guides/faststore/project-structure-config-options) file to optimize your store's SEO. These templates allow you to dynamically generate metadata based on the type of PLP, such as collections, categories, subcategories, or brands: 

Reviewer 2: ```suggestion
```

- Title template: This template will be applied to all PLPs, including collections, categories, subcategories, and brands.

> - Description template: This template is used for `collections`, as descriptions for `collections` are not configured in the Catalog. If a specific description isn't configured in the Catalog for a category, subcategory, or brand, this template will be used as a fallback.

Reviewer 2: Mover para depois do passo 2

> To understand the order to generate the SEO metadata for PLPs, see the [Understand how SEO metadata is determined](#understand-how-seo-metadata-is-determined) section.

Reviewer 2: ```suggestion
```

1. Open your FastStore project in a code editor of your preference.
2. Go to the `discovey.config.js` file and in the `seo` section add the following:

  ```discovery.config.js
  seo: {
    plp: {
      titleTemplate: '%s | FastStore PLP',
      descriptionTemplate: '%s products on FastStore Product Listing Page',
    },
  },
  ```

>   > ⚠ Change the values in `titleTemplate` and `descriptionTemplate` according to your needs. The `%s` should be used where you would like the PLP title/name to appear. 

Reviewer 2: ```suggestion
  > ⚠ Change the values in `titleTemplate` and `descriptionTemplate` according to your needs. Use %s as a placeholder for the PLP title/name.
```

### Step 2: Controlling lazy loading for custom sections

> By default, FastStore lazy-loads sections for performance optimization. However, certain custom sections (e.g., those containing H1 headers) may need to appear in the initial HTML for SEO purposes. To ensure these sections are included in the initial HTML, you can configure them to skip lazy loading.

Reviewer 2: ```suggestion
FastStore lazy-loads sections for performance reasons. However, certain [custom sections](https://developers.vtex.com/docs/guides/faststore/building-sections-creating-a-new-section), such as those containing H1 headers, may need to appear in the initial HTML for SEO purposes. To disable lazy loading for these sections, consider the following:
```

> Below are the steps to disable lazy loading for a custom section. In this example, we’ll use a custom section named `CustomPLPName`. If you haven’t created a custom section yet, refer to the [Creating a new section](https://developers.vtex.com/docs/guides/faststore/building-sections-creating-a-new-section) guide.

Reviewer 2: ```suggestion
```
